pool/pool_router.gno

```diff
@@ -0,0 +1,124 @@
+package pool
+
+import (
+	"strconv"
+	"strings"
+
+	"gno.land/p/demo/ufmt"
+)
+
+func FindBestPool(
+	tokenA string,
+	tokenB string,
+	zeroForOne bool,
+	amountSpecified bigint,
+) string {
+	if tokenA == tokenB {
+		panic("token pair cannot be the same")
+	}
+
+	if tokenA > tokenB {
+		tokenA, tokenB = tokenB, tokenA
+	}
+
+	partialPath := tokenA + "_" + tokenB + "_"
+	foundPool := []string{}
+
+	for poolPath, _ := range pools {
+		if strings.HasPrefix(poolPath, partialPath) {
+			foundPool = append(foundPool, poolPath)
+		}
+	}
+
+	// check if amount is enough
+	firstSelectedPool := []string{}
+	if false {
+		continue
+	} else if zeroForOne == true && amountSpecified > 0 {
+		for _, singlePool := range foundPool {
+			p := GetPoolFromPoolKey(singlePool)
+			poolSqrtX96 := p.slot0.sqrtPriceX96
+			poolPrice := poolSqrtX96 * poolSqrtX96 / Q96 / Q96
+
+			pool := GetPoolFromPoolKey(singlePool)
+			if pool.balances.token1 > amountSpecified*poolPrice { // must be bigger (can't be equal due to fee)
+				firstSelectedPool = append(firstSelectedPool, singlePool)
+			}
+		}
+
+	} else if zeroForOne == true && amountSpecified < 0 {
+		for _, singlePool := range foundPool {
+			pool := GetPoolFromPoolKey(singlePool)
+			if pool.balances.token1 > (-1 * amountSpecified) { // must be bigger (can't be equal due to fee)
+				firstSelectedPool = append(firstSelectedPool, singlePool)
+			}
+		}
+
+	} else if zeroForOne == false && amountSpecified > 0 {
+		for _, singlePool := range foundPool {
+			p := GetPoolFromPoolKey(singlePool)
+			poolSqrtX96 := p.slot0.sqrtPriceX96
+			poolPrice := poolSqrtX96 * poolSqrtX96 / Q96 / Q96
+
+			pool := GetPoolFromPoolKey(singlePool)
+			if pool.balances.token0 > amountSpecified/poolPrice { // must be bigger (can't be equal due to fee)
+				firstSelectedPool = append(firstSelectedPool, singlePool)
+			}
+		}
+
+	} else if zeroForOne == false && amountSpecified < 0 {
+		for _, singlePool := range foundPool {
+			pool := GetPoolFromPoolKey(singlePool)
+			if pool.balances.token0 > (-1 * amountSpecified) { // must be bigger (can't be equal due to fee)
```

Comment:
fee can't be to zero?

Reply:
it depends on swap situation, just wanted to include safe margin